## Cryptocurrency

In [1]:
import pandas as pd
import matplotlib.pyplot as plt

#### Load Data

In [2]:
crypto_data = pd.read_csv('./Resources/crypto_data.csv')

In [3]:
crypto_data

Unnamed: 0        CoinName Algorithm  IsTrading ProofType  \
0            42         42 Coin    Scrypt       True   PoW/PoS   
1           365         365Coin       X11       True   PoW/PoS   
2           404         404Coin    Scrypt       True   PoW/PoS   
3           611       SixEleven   SHA-256       True       PoW   
4           808             808   SHA-256       True   PoW/PoS   
...         ...             ...       ...        ...       ...   
1247        XBC     BitcoinPlus    Scrypt       True       PoS   
1248       DVTC      DivotyCoin    Scrypt      False   PoW/PoS   
1249       GIOT     Giotto Coin    Scrypt      False   PoW/PoS   
1250       OPSC  OpenSourceCoin   SHA-256      False   PoW/PoS   
1251       PUNK       SteamPunk       PoS      False       PoS   

      TotalCoinsMined TotalCoinSupply  
0        4.199995e+01              42  
1                 NaN      2300000000  
2        1.055185e+09       532000000  
3                 NaN          611000  
4        0.000000e+00               0  
...               ...             ...  
1247     1.283270e+05         1000000  
1248     2.149121e+07       100000000  
1249              NaN       233100000  
1250              NaN        21000000  
1251              NaN        40000000  

[1252 rows x 7 columns]

#### Data Preparation

Removing cryptocurrency that is not currently being traded.

In [4]:
crypto_1 = crypto_data.loc[(crypto_data["IsTrading"]==True)]
crypto_1

Unnamed: 0     CoinName    Algorithm  IsTrading ProofType  \
0            42      42 Coin       Scrypt       True   PoW/PoS   
1           365      365Coin          X11       True   PoW/PoS   
2           404      404Coin       Scrypt       True   PoW/PoS   
3           611    SixEleven      SHA-256       True       PoW   
4           808          808      SHA-256       True   PoW/PoS   
...         ...          ...          ...        ...       ...   
1243       SERO   Super Zero       Ethash       True       PoW   
1244        UOS          UOS      SHA-256       True      DPoI   
1245        BDX       Beldex  CryptoNight       True       PoW   
1246        ZEN      Horizen     Equihash       True       PoW   
1247        XBC  BitcoinPlus       Scrypt       True       PoS   

      TotalCoinsMined TotalCoinSupply  
0        4.199995e+01              42  
1                 NaN      2300000000  
2        1.055185e+09       532000000  
3                 NaN          611000  
4        0.000000e+00               0  
...               ...             ...  
1243              NaN      1000000000  
1244              NaN      1000000000  
1245     9.802226e+08      1400222610  
1246     7.296538e+06        21000000  
1247     1.283270e+05         1000000  

[1144 rows x 7 columns]

In [5]:
crypto_2 = crypto_1.drop(columns=['IsTrading'])
crypto_2

Unnamed: 0     CoinName    Algorithm ProofType  TotalCoinsMined  \
0            42      42 Coin       Scrypt   PoW/PoS     4.199995e+01   
1           365      365Coin          X11   PoW/PoS              NaN   
2           404      404Coin       Scrypt   PoW/PoS     1.055185e+09   
3           611    SixEleven      SHA-256       PoW              NaN   
4           808          808      SHA-256   PoW/PoS     0.000000e+00   
...         ...          ...          ...       ...              ...   
1243       SERO   Super Zero       Ethash       PoW              NaN   
1244        UOS          UOS      SHA-256      DPoI              NaN   
1245        BDX       Beldex  CryptoNight       PoW     9.802226e+08   
1246        ZEN      Horizen     Equihash       PoW     7.296538e+06   
1247        XBC  BitcoinPlus       Scrypt       PoS     1.283270e+05   

     TotalCoinSupply  
0                 42  
1         2300000000  
2          532000000  
3             611000  
4                  0  
...              ...  
1243      1000000000  
1244      1000000000  
1245      1400222610  
1246        21000000  
1247         1000000  

[1144 rows x 6 columns]

Droping rows with at least one missing element.

In [6]:
crypto_3 = crypto_2.dropna()
crypto_3

Unnamed: 0     CoinName    Algorithm ProofType  TotalCoinsMined  \
0            42      42 Coin       Scrypt   PoW/PoS     4.199995e+01   
2           404      404Coin       Scrypt   PoW/PoS     1.055185e+09   
4           808          808      SHA-256   PoW/PoS     0.000000e+00   
5          1337    EliteCoin          X13   PoW/PoS     2.927942e+10   
7           BTC      Bitcoin      SHA-256       PoW     1.792718e+07   
...         ...          ...          ...       ...              ...   
1238       ZEPH       ZEPHYR      SHA-256      DPoS     2.000000e+09   
1242        GAP      Gapcoin       Scrypt   PoW/PoS     1.493105e+07   
1245        BDX       Beldex  CryptoNight       PoW     9.802226e+08   
1246        ZEN      Horizen     Equihash       PoW     7.296538e+06   
1247        XBC  BitcoinPlus       Scrypt       PoS     1.283270e+05   

     TotalCoinSupply  
0                 42  
2          532000000  
4                  0  
5       314159265359  
7           21000000  
...              ...  
1238      2000000000  
1242       250000000  
1245      1400222610  
1246        21000000  
1247         1000000  

[685 rows x 6 columns]

Removing Cryptocurrency that has not been mined.

In [7]:
crypto_4 = crypto_3.loc[(crypto_3["TotalCoinsMined"]>0)]
crypto_4

Unnamed: 0     CoinName    Algorithm ProofType  TotalCoinsMined  \
0            42      42 Coin       Scrypt   PoW/PoS     4.199995e+01   
2           404      404Coin       Scrypt   PoW/PoS     1.055185e+09   
5          1337    EliteCoin          X13   PoW/PoS     2.927942e+10   
7           BTC      Bitcoin      SHA-256       PoW     1.792718e+07   
8           ETH     Ethereum       Ethash       PoW     1.076842e+08   
...         ...          ...          ...       ...              ...   
1238       ZEPH       ZEPHYR      SHA-256      DPoS     2.000000e+09   
1242        GAP      Gapcoin       Scrypt   PoW/PoS     1.493105e+07   
1245        BDX       Beldex  CryptoNight       PoW     9.802226e+08   
1246        ZEN      Horizen     Equihash       PoW     7.296538e+06   
1247        XBC  BitcoinPlus       Scrypt       PoS     1.283270e+05   

     TotalCoinSupply  
0                 42  
2          532000000  
5       314159265359  
7           21000000  
8                  0  
...              ...  
1238      2000000000  
1242       250000000  
1245      1400222610  
1246        21000000  
1247         1000000  

[532 rows x 6 columns]

Deleting CoinName as the data must be numeric to be comprehensible to a machine learning algorithm and it does not contribute to analysis.

In [8]:
crypto_5 = crypto_4.drop(columns=['Unnamed: 0','CoinName'])
crypto_5

Algorithm ProofType  TotalCoinsMined TotalCoinSupply
0          Scrypt   PoW/PoS     4.199995e+01              42
2          Scrypt   PoW/PoS     1.055185e+09       532000000
5             X13   PoW/PoS     2.927942e+10    314159265359
7         SHA-256       PoW     1.792718e+07        21000000
8          Ethash       PoW     1.076842e+08               0
...           ...       ...              ...             ...
1238      SHA-256      DPoS     2.000000e+09      2000000000
1242       Scrypt   PoW/PoS     1.493105e+07       250000000
1245  CryptoNight       PoW     9.802226e+08      1400222610
1246     Equihash       PoW     7.296538e+06        21000000
1247       Scrypt       PoS     1.283270e+05         1000000

[532 rows x 4 columns]

Using get_dummies() to convert features with text to numeric values. We have 5 columns and 532 rows.

In [9]:
crypto_6 = pd.get_dummies(crypto_5, columns=["Algorithm", "ProofType"])
crypto_6

TotalCoinsMined TotalCoinSupply  Algorithm_1GB AES Pattern Search  \
0        4.199995e+01              42                                 0   
2        1.055185e+09       532000000                                 0   
5        2.927942e+10    314159265359                                 0   
7        1.792718e+07        21000000                                 0   
8        1.076842e+08               0                                 0   
...               ...             ...                               ...   
1238     2.000000e+09      2000000000                                 0   
1242     1.493105e+07       250000000                                 0   
1245     9.802226e+08      1400222610                                 0   
1246     7.296538e+06        21000000                                 0   
1247     1.283270e+05         1000000                                 0   

      Algorithm_536  Algorithm_Argon2d  Algorithm_BLAKE256  Algorithm_Blake  \
0                 0                  0                   0                0   
2                 0                  0                   0                0   
5                 0                  0                   0                0   
7                 0                  0                   0                0   
8                 0                  0                   0                0   
...             ...                ...                 ...              ...   
1238              0                  0                   0                0   
1242              0                  0                   0                0   
1245              0                  0                   0                0   
1246              0                  0                   0                0   
1247              0                  0                   0                0   

      Algorithm_Blake2S  Algorithm_Blake2b  Algorithm_C11  ...  \
0                     0                  0              0  ...   
2                     0                  0              0  ...   
5                     0                  0              0  ...   
7                     0                  0              0  ...   
8                     0                  0              0  ...   
...                 ...                ...            ...  ...   
1238                  0                  0              0  ...   
1242                  0                  0              0  ...   
1245                  0                  0              0  ...   
1246                  0                  0              0  ...   
1247                  0                  0              0  ...   

      ProofType_PoW/PoS  ProofType_PoW/PoS   ProofType_PoW/PoW  \
0                     1                   0                  0   
2                     1                   0                  0   
5                     1                   0                  0   
7                     0                   0                  0   
8                     0                   0                  0   
...                 ...                 ...                ...   
1238                  0                   0                  0   
1242                  1                   0                  0   
1245                  0                   0                  0   
1246                  0                   0                  0   
1247                  0                   0                  0   

      ProofType_PoW/nPoS  ProofType_Pos  ProofType_Proof of Authority  \
0                      0              0                             0   
2                      0              0                             0   
5                      0              0                             0   
7                      0              0                             0   
8                      0              0                             0   
...                  ...            ...                           ...   
1238                   0              0                    

Wow! The number of columns increased by 94!

Standardize the dataset using StandardScaler to scale the data to have a mean of 0 and a variance of 1.

In [10]:
crypto_6.columns

Index(['TotalCoinsMined', 'TotalCoinSupply',
       'Algorithm_1GB AES Pattern Search', 'Algorithm_536',
       'Algorithm_Argon2d', 'Algorithm_BLAKE256', 'Algorithm_Blake',
       'Algorithm_Blake2S', 'Algorithm_Blake2b', 'Algorithm_C11',
       'Algorithm_Cloverhash', 'Algorithm_Counterparty',
       'Algorithm_CryptoNight', 'Algorithm_CryptoNight Heavy',
       'Algorithm_CryptoNight-V7', 'Algorithm_Cryptonight-GPU',
       'Algorithm_DPoS', 'Algorithm_Dagger', 'Algorithm_Dagger-Hashimoto',
       'Algorithm_ECC 256K1', 'Algorithm_Equihash',
       'Algorithm_Equihash+Scrypt', 'Algorithm_Ethash', 'Algorithm_Exosis',
       'Algorithm_Green Protocol', 'Algorithm_Groestl', 'Algorithm_HMQ1725',
       'Algorithm_HybridScryptHash256', 'Algorithm_IMesh',
       'Algorithm_Jump Consistent Hash', 'Algorithm_Keccak',
       'Algorithm_Leased POS', 'Algorithm_Lyra2RE', 'Algorithm_Lyra2REv2',
       'Algorithm_Lyra2Z', 'Algorithm_M7 POW', 'Algorithm_Multiple',
       'Algorithm_NIST5', 'Algor

In [11]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaled_data = scaler.fit_transform(crypto_6[["TotalCoinsMined", "TotalCoinSupply"]])
scaled_data = pd.DataFrame(scaled_data, columns=crypto_6.columns[:2])
scaled_data

TotalCoinsMined  TotalCoinSupply
0          -0.117108        -0.152870
1          -0.093970        -0.145009
2           0.524946         4.489424
3          -0.116715        -0.152560
4          -0.114747        -0.152870
..               ...              ...
527        -0.073251        -0.123317
528        -0.116781        -0.149176
529        -0.095613        -0.132179
530        -0.116948        -0.152560
531        -0.117105        -0.152856

[532 rows x 2 columns]

In [12]:
crypto_7 = crypto_6[crypto_6.columns[2:]]
crypto_7 = crypto_7.reset_index()
crypto_7

index  Algorithm_1GB AES Pattern Search  Algorithm_536  \
0        0                                 0              0   
1        2                                 0              0   
2        5                                 0              0   
3        7                                 0              0   
4        8                                 0              0   
..     ...                               ...            ...   
527   1238                                 0              0   
528   1242                                 0              0   
529   1245                                 0              0   
530   1246                                 0              0   
531   1247                                 0              0   

     Algorithm_Argon2d  Algorithm_BLAKE256  Algorithm_Blake  \
0                    0                   0                0   
1                    0                   0                0   
2                    0                   0                0   
3                    0                   0                0   
4                    0                   0                0   
..                 ...                 ...              ...   
527                  0                   0                0   
528                  0                   0                0   
529                  0                   0                0   
530                  0                   0                0   
531                  0                   0                0   

     Algorithm_Blake2S  Algorithm_Blake2b  Algorithm_C11  \
0                    0                  0              0   
1                    0                  0              0   
2                    0                  0              0   
3                    0                  0              0   
4                    0                  0              0   
..                 ...                ...            ...   
527                  0                  0              0   
528                  0                  0              0   
529                  0                  0              0   
530                  0                  0              0   
531                  0                  0              0   

     Algorithm_Cloverhash  ...  ProofType_PoW/PoS  ProofType_PoW/PoS   \
0                       0  ...                  1                   0   
1                       0  ...                  1                   0   
2                       0  ...                  1                   0   
3                       0  ...                  0                   0   
4                       0  ...                  0                   0   
..                    ...  ...                ...                 ...   
527                     0  ...                  0                   0   
528                     0  ...                  1                   0   
529                     0  ...                  0                   0   
530                     0  ...                  0                   0   
531                     0  ...                  0                   0   

     ProofType_PoW/PoW  ProofType_PoW/nPoS  ProofType_Pos  \
0                    0                   0              0   
1                    0                   0              0   
2                    0                   0              0   
3                    0                   0              0   
4                    0                   0              0   
..                 ...                 ...            ...   
527                  0                   0              0   
528                  0                   0              0   
529                  0                   0              0   
530                  0                   0              0   
531                  0                   0              0   

     ProofType_Proof of Authority  ProofType_Proof of Trust  ProofType_TPoS  \
0                               0                         0               0   
1 

In [13]:
crypto_8 = crypto_7.drop(columns=['index'])
crypto_8

Algorithm_1GB AES Pattern Search  Algorithm_536  Algorithm_Argon2d  \
0                                   0              0                  0   
1                                   0              0                  0   
2                                   0              0                  0   
3                                   0              0                  0   
4                                   0              0                  0   
..                                ...            ...                ...   
527                                 0              0                  0   
528                                 0              0                  0   
529                                 0              0                  0   
530                                 0              0                  0   
531                                 0              0                  0   

     Algorithm_BLAKE256  Algorithm_Blake  Algorithm_Blake2S  \
0                     0                0                  0   
1                     0                0                  0   
2                     0                0                  0   
3                     0                0                  0   
4                     0                0                  0   
..                  ...              ...                ...   
527                   0                0                  0   
528                   0                0                  0   
529                   0                0                  0   
530                   0                0                  0   
531                   0                0                  0   

     Algorithm_Blake2b  Algorithm_C11  Algorithm_Cloverhash  \
0                    0              0                     0   
1                    0              0                     0   
2                    0              0                     0   
3                    0              0                     0   
4                    0              0                     0   
..                 ...            ...                   ...   
527                  0              0                     0   
528                  0              0                     0   
529                  0              0                     0   
530                  0              0                     0   
531                  0              0                     0   

     Algorithm_Counterparty  ...  ProofType_PoW/PoS  ProofType_PoW/PoS   \
0                         0  ...                  1                   0   
1                         0  ...                  1                   0   
2                         0  ...                  1                   0   
3                         0  ...                  0                   0   
4                         0  ...                  0                   0   
..                      ...  ...                ...                 ...   
527                       0  ...                  0                   0   
528                       0  ...                  1                   0   
529                       0  ...                  0                   0   
530                       0  ...                  0                   0   
531                       0  ...                  0                   0   

     ProofType_PoW/PoW  ProofType_PoW/nPoS  ProofType_Pos  \
0                    0                   0              0   
1                    0                   0              0   
2                    0                   0              0   
3                    0                   0              0   
4                    0                   0              0   
..                 ...                 ...            ...   
527                  0                   0              0   
528                  0                   0              0   
529                  0                   0              0   
530                  0                   0              0   
531               

In [14]:
crypto_8["TotalCoinSupply"] = scaled_data["TotalCoinSupply"]
crypto_8

Algorithm_1GB AES Pattern Search  Algorithm_536  Algorithm_Argon2d  \
0                                   0              0                  0   
1                                   0              0                  0   
2                                   0              0                  0   
3                                   0              0                  0   
4                                   0              0                  0   
..                                ...            ...                ...   
527                                 0              0                  0   
528                                 0              0                  0   
529                                 0              0                  0   
530                                 0              0                  0   
531                                 0              0                  0   

     Algorithm_BLAKE256  Algorithm_Blake  Algorithm_Blake2S  \
0                     0                0                  0   
1                     0                0                  0   
2                     0                0                  0   
3                     0                0                  0   
4                     0                0                  0   
..                  ...              ...                ...   
527                   0                0                  0   
528                   0                0                  0   
529                   0                0                  0   
530                   0                0                  0   
531                   0                0                  0   

     Algorithm_Blake2b  Algorithm_C11  Algorithm_Cloverhash  \
0                    0              0                     0   
1                    0              0                     0   
2                    0              0                     0   
3                    0              0                     0   
4                    0              0                     0   
..                 ...            ...                   ...   
527                  0              0                     0   
528                  0              0                     0   
529                  0              0                     0   
530                  0              0                     0   
531                  0              0                     0   

     Algorithm_Counterparty  ...  ProofType_PoW/PoS   ProofType_PoW/PoW  \
0                         0  ...                   0                  0   
1                         0  ...                   0                  0   
2                         0  ...                   0                  0   
3                         0  ...                   0                  0   
4                         0  ...                   0                  0   
..                      ...  ...                 ...                ...   
527                       0  ...                   0                  0   
528                       0  ...                   0                  0   
529                       0  ...                   0                  0   
530                       0  ...                   0                  0   
531                       0  ...                   0                  0   

     ProofType_PoW/nPoS  ProofType_Pos  ProofType_Proof of Authority  \
0                     0              0                             0   
1                     0              0                             0   
2                     0              0                             0   
3                     0              0                             0   
4                     0              0                             0   
..                  ...            ...                           ...   
527                   0              0                             0   
528                   0              0                             0   
529                   0              0   

In [15]:
crypto_8["TotalCoinsMined"] = scaled_data["TotalCoinsMined"]
crypto_8

Algorithm_1GB AES Pattern Search  Algorithm_536  Algorithm_Argon2d  \
0                                   0              0                  0   
1                                   0              0                  0   
2                                   0              0                  0   
3                                   0              0                  0   
4                                   0              0                  0   
..                                ...            ...                ...   
527                                 0              0                  0   
528                                 0              0                  0   
529                                 0              0                  0   
530                                 0              0                  0   
531                                 0              0                  0   

     Algorithm_BLAKE256  Algorithm_Blake  Algorithm_Blake2S  \
0                     0                0                  0   
1                     0                0                  0   
2                     0                0                  0   
3                     0                0                  0   
4                     0                0                  0   
..                  ...              ...                ...   
527                   0                0                  0   
528                   0                0                  0   
529                   0                0                  0   
530                   0                0                  0   
531                   0                0                  0   

     Algorithm_Blake2b  Algorithm_C11  Algorithm_Cloverhash  \
0                    0              0                     0   
1                    0              0                     0   
2                    0              0                     0   
3                    0              0                     0   
4                    0              0                     0   
..                 ...            ...                   ...   
527                  0              0                     0   
528                  0              0                     0   
529                  0              0                     0   
530                  0              0                     0   
531                  0              0                     0   

     Algorithm_Counterparty  ...  ProofType_PoW/PoW  ProofType_PoW/nPoS  \
0                         0  ...                  0                   0   
1                         0  ...                  0                   0   
2                         0  ...                  0                   0   
3                         0  ...                  0                   0   
4                         0  ...                  0                   0   
..                      ...  ...                ...                 ...   
527                       0  ...                  0                   0   
528                       0  ...                  0                   0   
529                       0  ...                  0                   0   
530                       0  ...                  0                   0   
531                       0  ...                  0                   0   

     ProofType_Pos  ProofType_Proof of Authority  ProofType_Proof of Trust  \
0                0                             0                         0   
1                0                             0                         0   
2                0                             0                         0   
3                0                             0                         0   
4                0                             0                         0   
..             ...                           ...                       ...   
527              0                             0                         0   
528              0                             0                 

### Dimensionality Reduction

In [18]:
from sklearn.decomposition import PCA

# Initialize PCA model
pca = PCA(n_components=0.90)

# Get components for the data.
crypto_pca = pca.fit_transform(crypto_8)
crypto_pca

array([[-0.20927594, -0.80176522,  0.17297514, ..., -0.02304258,
        -0.00784955,  0.01578897],
       [-0.18738959, -0.80425863,  0.16255518, ..., -0.0230814 ,
        -0.00775231,  0.01588806],
       [ 3.52967377, -0.30553695,  2.9154029 , ...,  0.36081392,
        -0.23380143, -0.07021991],
       ...,
       [-0.12458946,  0.73619802, -0.0904179 , ...,  0.4309651 ,
        -0.08535956,  0.07079805],
       [-0.17492407,  0.69566945, -0.15519312, ..., -0.21383758,
         0.12157764,  0.83296946],
       [-0.19826273, -0.23434647, -0.04325234, ...,  0.11206426,
         0.07759961,  0.06541967]])

In [19]:
crypto_pca[0]

array([-0.20927594, -0.80176522,  0.17297514, -0.6723849 , -0.02802675,
       -0.02734979,  0.00216075, -0.01685939, -0.04742258, -0.02304258,
       -0.00784955,  0.01578897])

Woah. The number of features were reduced from 98 to 12.

In [20]:
# Transform PCA data to a DataFrame
df_crypto_pca = pd.DataFrame(
    data=crypto_pca, columns=["principal component 1", "principal component 2", "principal component 3", 
                                "principal component 4", "principal component 5", "principal component 6", 
                                "principal component 7", "principal component 8", "principal component 9",
                                "principal component 10", "principal component 11", "principal component 12"]
)
df_crypto_pca.head()

principal component 1  principal component 2  principal component 3  \
0              -0.209276              -0.801765               0.172975   
1              -0.187390              -0.804259               0.162555   
2               3.529674              -0.305537               2.915403   
3              -0.180157               0.694290              -0.165771   
4              -0.173289               0.688569              -0.154390   

   principal component 4  principal component 5  principal component 6  \
0              -0.672385              -0.028027              -0.027350   
1              -0.673029              -0.028782              -0.027213   
2               0.414815              -0.163323              -0.259278   
3               0.239931              -0.278380              -0.614707   
4               0.131918              -0.212548              -0.104388   

   principal component 7  principal component 8  principal component 9  \
0               0.002161              -0.016859              -0.047423   
1               0.002265              -0.017852              -0.047565   
2              -0.278472              -0.152434               0.747395   
3               0.624124              -0.015063               0.027769   
4              -0.242103               0.002800               0.113068   

   principal component 10  principal component 11  principal component 12  
0               -0.023043               -0.007850                0.015789  
1               -0.023081               -0.007752                0.015888  
2                0.360814               -0.233801               -0.070220  
3                0.009259               -0.026164               -0.026461  
4               -0.125626                0.058157               -0.129333